Source/Evolve/WarpXComputeDt.cpp

```diff
@@ -91,10 +91,41 @@ WarpX::ComputeDt ()
 
     dt.resize(0);
     dt.resize(max_level+1,deltat);
+    dt_next.resize(0);
+    dt_next.resize(max_level+1,deltat);
 
     if (do_subcycling) {
         for (int lev = max_level-1; lev >= 0; --lev) {
             dt[lev] = dt[lev+1] * refRatio(lev)[0];
+            dt_next[lev] = dt_next[lev+1] * refRatio(lev)[0];
+        }
+    }
+}
+
+/**
+ * Determine the simulation timestep from the maximum speed of all particles
+ * Sets timestep so that a particle can only cross cfl*dx cells per timestep.
+ */
+void
+WarpX::UpdateDtFromParticleSpeeds ()
+{
+    const amrex::Real* dx = geom[max_level].CellSize();
+    amrex::Real dx_min = minDim(dx);
+
+    const amrex::ParticleReal max_v = mypc->maxParticleVelocity();
+    const amrex::Real deltat_new = cfl * dx_min / max_v;
+
+    // Set present dt to previous next dt
+    dt[max_level] = dt_next[max_level];
+    dt_next[max_level] = deltat_new;
+
+    for (int lev = max_level-1; lev >= 0; --lev) {
+        if (do_subcycling) {
+            dt[lev] = dt[lev+1] * refRatio(lev)[0];
```

Comment:
Could this be `dt[lev] = dt_next[lev];` and taken out of the `if` blocks?

Reply:
Yes, I think so. That would give 
```cpp
for (int lev = max_level-1; lev >= 0; --lev) {
    dt[lev] = dt_next[lev];
    if (do_subcycling) {
        dt_next[lev] = dt_next[lev+1] * refRatio(lev)[0];
    } else {
        dt_next[lev] = dt_next[lev+1];
    }
}
```